apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/keyboard-navigation.tsx

```diff
@@ -0,0 +1,51 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+/**
+ * Internal dependencies
+ */
+import useFocusHandler from './hooks/use-focus-handler';
+import useKeydownHandler from './hooks/use-keydown-handler';
+
+interface Props {
+	onMinimize: () => void;
+	onDismiss: ( target: string ) => () => void;
+	onNextCardProgression: () => void;
+	onPreviousCardProgression: () => void;
+	focusRef: React.MutableRefObject< null | HTMLElement >;
+	isMinimized: boolean;
+}
+
+const KeyboardNavigation: React.FunctionComponent< Props > = ( {
+	onMinimize,
+	onDismiss,
+	onNextCardProgression,
+	onPreviousCardProgression,
+	focusRef,
+	isMinimized,
+} ) => {
+	function ExpandedTourNav() {
+		useKeydownHandler( {
+			onEscape: onMinimize,
+			onArrowRight: onNextCardProgression,
+			onArrowLeft: onPreviousCardProgression,
+		} );
+		return null;
+	}
+
+	function MinimizedTourNav() {
+		useKeydownHandler( { onEscape: onDismiss( 'esc-key' ) } );
```

Comment:
It's just an idea and a matter of taste, but I think we can condense these components into one by using conditional properties (example [here](https://stackoverflow.com/questions/11704267/in-javascript-how-to-conditionally-add-a-member-to-an-object)). 

Something like this: 

```js
function TourNav() {
		useKeydownHandler( {
			...( isMinimized && {
				onArrowRight: onNextCardProgression,
				onArrowLeft: onPreviousCardProgression,
			} ),
			onEscape: isMinimized ? onDismiss( 'esc-key' ) : onMinimize,
		} );

		return null;
	}
```

Reply:
It's not a bad idea Stefan. I think there are just drawbacks, as with everything else. This is obviously a minor case, and no bad design can really complicate things. Irrespective, I think it's a good idea to ask whenever we want to introduce some abstraction whether it justifies the means: does it introduce complexity, does it make things easier/harder to follow, is it all justified, etc. 

In our case, we seem to have one layer of abstraction already (`KeyboardNavigation`) that's akin to a factory method. It has the form: `KeyboardNavigation: { ExpandedTourNav, MinimizedTourNav }`. (none of these exist outside of the "tour" btw). With `TourNav` above, we effectively introduce another layer (`ExpandedOrMinimizedTourNav`) to make the two classes "condensed". 

A condensed form sounds like a neat idea on the surface with cool conditional properties and the like. Is it really ideal though? It does introduce some complexity right? - there are ternaries and conditional object spreading (I need to take a moment and think about how that object is constructed), the decision logic is also kinda separated now from the main flow, concerns are somewhat generalized too in this condensed form. What happens if we introduce more to the mix - more views maybe (if such a thing is possible)? We'll keep adding to the condensed form, handle more conditions in those expressions, then start separating things into variables, and I think we'll probably navigate back to separating concerns a layer below (so where we start). 🤔 

Drawbacks drawbacks drawbacks. Everything has drawbacks. The current approach has drawbacks.

Reply:
> With TourNav above, we effectively introduce another layer (ExpandedOrMinimizedTourNav) to make the two classes "condensed".

We don't introduce another layer but we replace `KeyboardNavigation: { ExpandedTourNav, MinimizedTourNav }` with `KeyboardNavigation: { ExpandedOrMinimizedTourNav}`. So we're simplifying the signature by reducing the parameters.

> Drawbacks drawbacks drawbacks. Everything has drawbacks.

I'm quite fond of the DRY principle, since, in general, it makes code better maintainable, and often more comprehensible, but not in all cases, and I believe that this case is the exception to the rule. Condensing both layers into one will make the code more DRY, but less readable and (probably) less maintainable. Especially when potentially more views are added in the future. 

So, I think that both solutions are fine for the current situation, but if we take the future into account, you convinced me that it's better to keep the current solution.

Reply:
> We don't introduce another layer but we replace KeyboardNavigation: { ExpandedTourNav, MinimizedTourNav } with KeyboardNavigation: { ExpandedOrMinimizedTourNav}. So we're simplifying the signature by reducing the parameters.

I think we most certainly are introducing an abstraction with `ExpandedOrMinimizedTourNav`.

> I'm quite fond of the DRY principle, since, in general, it makes code better maintainable, and often more comprehensible, but not in all cases, and I believe that this case is the exception to the rule.

DRY is great! I think one of the commonest mistakes that people make in reference to DRY is treating it as another way of saying "don't copy/paste lines of code". This is the tiniest, most trivial part about DRY IMO. To paraphrase a [favourite book](https://pragprog.com/titles/tpp20/the-pragmatic-programmer-20th-anniversary-edition/) on the subject: 

> DRY is about the duplication of _knowledge_, of _intent_. It’s about expressing the same thing in two different places, possibly in two totally different ways.
> 
> Here’s the acid test: when some single facet of the code has to change, do you find yourself making that change in multiple places, and in multiple different formats? Do you have to change code and documentation, or a database schema and a structure that holds it, or...? If so, your code isn’t DRY.

Let's not take that to mean every single condition and function used in a codebase should be abstracted behind facades and complex interfaces. [Code simplicity](p1bHI1-yn-p2) is such a great, fundamental rule in software development that pretty much every other pattern/design/approach should feed into it one way or another. There are many great principles to refer to. The challenge is understanding their context and applicability.